In [1]:
import pandas as pd

In [3]:
df = pd.read_csv(r'''C:\Users\MiaZo_000\Desktop\LoanData.csv''',skiprows=1)

C:\Users\MiaZo_000\Anaconda3\lib\site-packages\IPython\core\interactiveshell.py:2698: DtypeWarning: Columns (0,47) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 42542 entries, 0 to 42541
Columns: 145 entries, id to settlement_term
dtypes: float64(115), object(30)
memory usage: 47.1+ MB


In [6]:
df.select_dtypes(exclude='O').describe().T

count          mean  \
member_id                                       0.0           NaN   
loan_amnt                                   42535.0  11089.722581   
funded_amnt                                 42535.0  10821.585753   
funded_amnt_inv                             42535.0  10139.938785   
installment                                 42535.0    322.623063   
annual_inc                                  42531.0  69136.556420   
url                                             0.0           NaN   
dti                                         42535.0     13.373043   
delinq_2yrs                                 42506.0      0.152449   
inq_last_6mths                              42506.0      1.081424   
mths_since_last_delinq                      15609.0     35.017618   
mths_since_last_record                       3651.0     59.175568   
open_acc                                    42506.0      9.343951   
pub_rec                                     42506.0      0.058156   
revol_bal                                   42535.0  14297.860915   
total_acc                                   42506.0     22.124406   
out_prncp                                   42535.0      0.000000   
out_prncp_inv                               42535.0      0.000000   
total_pymnt                                 42535.0  12020.265657   
total_pymnt_inv                             42535.0  11312.805189   
total_rec_prncp                             42535.0   9675.675936   
total_rec_int                               42535.0   2239.981444   
total_rec_late_fee                          42535.0      1.516890   
recoveries                                  42535.0    103.091480   
collection_recovery_fee                     42535.0     14.401525   
last_pymnt_amnt                             42535.0   2613.245652   
collections_12_mths_ex_med                  42390.0      0.000000   
mths_since_last_major_derog                     0.0           NaN   
policy_code                                 42535.0      1.000000   
annual_inc_joint                                0.0           NaN   
...                                             ...           ...   
total_bc_limit                                  0.0           NaN   
total_il_high_credit_limit                      0.0           NaN   
revol_bal_joint                                 0.0           NaN   
sec_app_earliest_cr_line                        0.0           NaN   
sec_app_inq_last_6mths                          0.0           NaN   
sec_app_mort_acc                                0.0           NaN   
sec_app_open_acc                                0.0           NaN   
sec_app_revol_util                              0.0           NaN   
sec_app_open_act_il                             0.0           NaN   
sec_app_num_rev_accts                           0.0           NaN   
sec_app_chargeoff_within_12_mths                0.0           NaN   
sec_app_collections_12_mths_ex_med              0.0           NaN   
sec_app_mths_since_last_major_derog             0.0           NaN   
hardship_type                                   0.0           NaN   
hardship_reason                                 0.0           NaN   
hardship_status                                 0.0           NaN   
deferral_term                                   0.0           NaN   
hardship_amount                                 0.0           NaN   
hardship_start_date                             0.0           NaN   
hardship_end_date                               0.0           NaN   
payment_plan_start_date                         0.0           NaN   
hardship_length                                 0.0           NaN   
hardship_dpd                                    0.0           NaN   
hardship_loan_status                            0.0           NaN   
orig_projected_additional_accrued_interest      0.0           NaN   
hardship_payoff_balance_amount                  0.0           NaN   
hardship_last_payment_amount                

There are total 42542 rows and quite a few features does not have a non-null values at all. I am going to delete those columns that have too many nulls.

In [8]:
df = df.dropna(thresh=len(df)*0.5,axis=1)

In [11]:
df.select_dtypes(include='O').describe().T

count unique                 top   freq
term                  42535      2           36 months  31534
int_rate              42535    394              10.99%    970
grade                 42535      7                   B  12389
sub_grade             42535     35                  B3   2997
emp_title             39909  30658             US Army    139
emp_length            41423     11           10+ years   9369
home_ownership        42535      5                RENT  20181
verification_status   42535      3        Not Verified  18758
issue_d               42535     55              Dec-11   2267
loan_status           42535      4          Fully Paid  34116
pymnt_plan            42535      1                   n  42535
desc                  29242  28963                        225
purpose               42535     14  debt_consolidation  19776
title                 42522  21256  Debt Consolidation   2259
zip_code              42535    837               100xx    649
addr_state            42535     50                  CA   7429
earliest_cr_line      42506    530              Oct-99    393
revol_util            42445   1119                  0%   1070
initial_list_status   42535      1                   f  42535
last_pymnt_d          42452    112              Mar-13   1070
last_credit_pull_d    42531    135              Oct-18   9087
application_type      42535      1          Individual  42535
hardship_flag         42535      1                   N  42535
disbursement_method   42535      1                Cash  42535
debt_settlement_flag  42535      2                   N  42375

Notice there are some features only having one value. Those should be dropped as well.

In [13]:
# Drop columns that have one unique value
df = df.loc[:,df.apply(lambda x:x.nunique() != 1)]
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 42542 entries, 0 to 42541
Data columns (total 44 columns):
loan_amnt                  42535 non-null float64
funded_amnt                42535 non-null float64
funded_amnt_inv            42535 non-null float64
term                       42535 non-null object
int_rate                   42535 non-null object
installment                42535 non-null float64
grade                      42535 non-null object
sub_grade                  42535 non-null object
emp_title                  39909 non-null object
emp_length                 41423 non-null object
home_ownership             42535 non-null object
annual_inc                 42531 non-null float64
verification_status        42535 non-null object
issue_d                    42535 non-null object
loan_status                42535 non-null object
desc                       29242 non-null object
purpose                    42535 non-null object
title                      42522 non-null object
zip_

#### Now, the number of columns is cut down from 145 to 44. I am going to look at the meta data and to understand what those features are. This loan data file contains the complete information of the loans, including the current status of loans and last payment, so I need to drop those variables that only make sense after loans are approved.

### I need to go back to the meta data file with the remaining variables. I copied the list of columns above into excel and used the function, VLOOKUP, to find the explanations.

![image](/image.png)
